# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/style/FontFaceSet.cpp
# Commit: ae329eaf22f7
# Full Hash: ae329eaf22f70199e5f3f9d16d4c3b7c50120732
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2022-06-29 03:44:30
# Regressor Bug: 1771493
# File Overlap Count: 1
# Description:
#   Bug 1771493 - Part 4. Split FontFaceSetDocumentImpl from FontFaceSetImpl. r=emilio
#   
#   This patch splits document specific functionality from FontFaceSetImpl
#   into a new class FontFaceSetDocumentImpl. This will make it easier to
#   fork FontFaceSetImpl for workers.
# ==============================================================================

diff -r f724c21b1f15 -r ae329eaf22f7 layout/style/FontFaceSet.cpp
--- a/layout/style/FontFaceSet.cpp	Tue Jun 28 21:47:00 2022 +0000
+++ b/layout/style/FontFaceSet.cpp	Tue Jun 28 21:47:01 2022 +0000
@@ -16,6 +16,7 @@
 #include "mozilla/dom/Event.h"
 #include "mozilla/dom/FontFaceImpl.h"
 #include "mozilla/dom/FontFaceSetBinding.h"
+#include "mozilla/dom/FontFaceSetDocumentImpl.h"
 #include "mozilla/dom/FontFaceSetIterator.h"
 #include "mozilla/dom/FontFaceSetLoadEvent.h"
 #include "mozilla/dom/FontFaceSetLoadEventBinding.h"
@@ -93,11 +94,8 @@
 NS_INTERFACE_MAP_BEGIN_CYCLE_COLLECTION(FontFaceSet)
 NS_INTERFACE_MAP_END_INHERITING(DOMEventTargetHelper)
 
-FontFaceSet::FontFaceSet(nsPIDOMWindowInner* aWindow, dom::Document* aDocument)
-    : DOMEventTargetHelper(aWindow),
-      mImpl(new FontFaceSetImpl(this, aDocument)) {
-  mImpl->Initialize();
-}
+FontFaceSet::FontFaceSet(nsIGlobalObject* aParent)
+    : DOMEventTargetHelper(aParent) {}
 
 FontFaceSet::~FontFaceSet() {
   // Assert that we don't drop any FontFaceSet objects during a Servo traversal,
@@ -107,6 +105,16 @@
   Destroy();
 }
 
+/* static */ already_AddRefed<FontFaceSet> FontFaceSet::CreateForDocument(
+    dom::Document* aDocument) {
+  RefPtr<FontFaceSet> set = new FontFaceSet(aDocument->GetScopeObject());
+  RefPtr<FontFaceSetDocumentImpl> impl =
+      new FontFaceSetDocumentImpl(set, aDocument);
+  impl->Initialize();
+  set->mImpl = std::move(impl);
+  return set.forget();
+}
+
 JSObject* FontFaceSet::WrapObject(JSContext* aContext,
                                   JS::Handle<JSObject*> aGivenProto) {
   return FontFaceSet_Binding::Wrap(aContext, this, aGivenProto);